### ISSUE TITLE
When using horovod on two GPUs  of a machine, the backward time is twice as long as that on a single GPU

### ISSUE THREAD

**Issue by GG-Bond-007** [horovod/horovod#2828]
I use horovod in docker horovod/horovod:0.20.0-tf2.3.0-torch1.6.0-mxnet1.6.0.post0-py3.7-cuda10.1.
`def train(train_loader, model, epoch, optimizer, datasets):
    epoch_time = utils.AverageMeter('Time', ':6.3f')
    data_time = utils.AverageMeter('Data', ':6.3f')
    backward_time = utils.AverageMeter('backward', ':6.3f')
    losses = utils.AverageMeter('Loss', ':6.4f')
    progress = utils.ProgressMeter(len(train_loader),[epoch_time, data_time,losses,backward_time], prefix="Epoch: [{}]".format(epoch))
    end_data_time = time.time()
    end_epoch_time = time.time()

    # for batch_idx, (im_1, im_2) in enumerate(train_loader):
    for batch_idx, (images, target) in enumerate(train_loader):
        im_1 = images[0]
        im_2 = images[1]
        im_1 = im_1.cuda()
        im_2 = im_2.cuda()
        data_time.update(time.time() - end_data_time)

        loss = model(im_1, im_2)
        losses.update(loss.item(), im_1.size(0))

        backward_time_start = time.time()
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        backward_time.update(time.time() - backward_time_start)

        end_data_time = time.time()
        if batch_idx % 50 == 0:
            epoch_time.update(time.time() - end_epoch_time)
            end_epoch_time = time.time()
            progress.display(batch_idx)

    # save model
    if hvd.rank() == 0:
        model_file_name = './' + datasets + '/train_model_horovod.pth'
        torch.save(model.state_dict(), model_file_name)

hvd.init()
torch.cuda.set_device(hvd.local_rank())

# hyperparameters
base_encoder = 'resnet50'
dim = 2048
epochs = 100
batchsize = 32
datasets = 'ImageNet'

# build model
model = simsiam.SimSiam(base_encoder, dim)
model.cuda()
model.train()
cudnn.benchmark = True

# Data loading code
normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                 std=[0.229, 0.224, 0.225])
augmentation = [
    transforms.RandomResizedCrop(224, scale=(0.2, 1.)),
    transforms.RandomApply([
        transforms.ColorJitter(0.4, 0.4, 0.4, 0.1)  # not strengthened
    ], p=0.8),
    transforms.RandomGrayscale(p=0.2),
    transforms.RandomApply([utils.GaussianBlur([.1, 2.])], p=0.5),
    transforms.RandomHorizontalFlip(),
    transforms.RandomGrayscale(p=0.2),
    transforms.ToTensor(),
    normalize
]

train_dataset = torchvision.datasets.folder.ImageFolder(root='/ImageNet/train', transform=utils.TwoCropsTransform(transforms.Compose(augmentation)))
train_sampler = torch.utils.data.distributed.DistributedSampler(train_dataset, num_replicas=hvd.size(), rank=hvd.rank())
train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=batchsize, num_workers=8, sampler=train_sampler,
                                           pin_memory=True, drop_last=True)
print('Data loading is finished')

# initial the hvd
optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9, weight_decay=0.0001)
hvd.broadcast_parameters(model.state_dict(), root_rank=0)
compression = hvd.Compression.fp16
optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=model.named_parameters(),compression=compression)
hvd.broadcast_optimizer_state(optimizer, root_rank=0)
print('start to train')

for epoch in range(epochs):
    train_sampler.set_epoch(epoch)
    utils.adjust_learning_rate(optimizer, epoch, batchsize*4)
    train(train_loader, model, epoch, optimizer, datasets)`

The machine has 2 GPUs,when I run the code using horovodrun -np 2 --start-timeout 300 python run.py,  the output is 
[1,1]<stdout>:Epoch: [0][    0/20018]	Time  8.448 ( 8.448)	Data  6.697 ( 6.697)	Loss 0.0008 (0.0008)	backward  1.043 ( 1.043)
[1,0]<stdout>:Epoch: [0][    0/20018]	Time  8.462 ( 8.462)	Data  4.035 ( 4.035)	Loss 0.0005 (0.0005)	backward  3.777 ( 3.777)
[1,0]<stdout>:Epoch: [0][   50/20018]	Time 33.566 (21.014)	Data  0.019 ( 0.108)	Loss -0.0430 (-0.0190)	backward  0.463 ( 0.502)
[1,1]<stdout>:Epoch: [0][   50/20018]	Time 33.604 (21.026)	Data  0.034 ( 0.159)	Loss -0.0464 (-0.0184)	backward  0.297 ( 0.458)
[1,0]<stdout>:Epoch: [0][  100/20018]	Time 32.605 (24.878)	Data  0.063 ( 0.067)	Loss -0.0633 (-0.0365)	backward  0.322 ( 0.471)
[1,1]<stdout>:Epoch: [0][  100/20018]	Time 32.600 (24.884)	Data  0.037 ( 0.091)	Loss -0.0710 (-0.0360)	backward  0.345 ( 0.441)
[1,0]<stdout>:Epoch: [0][  150/20018]	Time 32.910 (26.886)	Data  0.010 ( 0.053)	Loss -0.0828 (-0.0507)	backward  0.639 ( 0.466)
[1,1]<stdout>:Epoch: [0][  150/20018]	Time 32.894 (26.886)	Data  0.014 ( 0.070)	Loss -0.1050 (-0.0509)	backward  0.462 ( 0.432)
[1,0]<stdout>:Epoch: [0][  200/20018]	Time 33.171 (28.143)	Data  0.005 ( 0.046)	Loss -0.0710 (-0.0648)	backward  0.478 ( 0.458)
[1,1]<stdout>:Epoch: [0][  200/20018]	Time 33.180 (28.145)	Data  0.009 ( 0.060)	Loss -0.1339 (-0.0647)	backward  0.383 ( 0.429)
[1,1]<stdout>:Epoch: [0][  250/20018]	Time 33.498 (29.037)	Data  0.045 ( 0.053)	Loss -0.1771 (-0.0785)	backward  0.305 ( 0.433)
[1,0]<stdout>:Epoch: [0][  250/20018]	Time 33.529 (29.040)	Data  0.025 ( 0.042)	Loss -0.1311 (-0.0768)	backward  0.361 ( 0.451)
[1,0]<stdout>:Epoch: [0][  300/20018]	Time 32.773 (29.574)	Data  0.032 ( 0.040)	Loss -0.1738 (-0.0894)	backward  0.456 ( 0.448)
[1,1]<stdout>:Epoch: [0][  300/20018]	Time 32.797 (29.574)	Data  0.007 ( 0.049)	Loss -0.1685 (-0.0919)	backward  0.458 ( 0.431)
[1,1]<stdout>:Epoch: [0][  350/20018]	Time 33.129 (30.019)	Data  0.020 ( 0.045)	Loss -0.1833 (-0.1047)	backward  0.390 ( 0.434)
[1,0]<stdout>:Epoch: [0][  350/20018]	Time 33.151 (30.021)	Data  0.007 ( 0.039)	Loss -0.1780 (-0.1023)	backward  0.489 ( 0.445)
[1,0]<stdout>:Epoch: [0][  400/20018]	Time 33.130 (30.366)	Data  0.041 ( 0.037)	Loss -0.2152 (-0.1147)	backward  0.504 ( 0.446)
[1,1]<stdout>:Epoch: [0][  400/20018]	Time 33.150 (30.367)	Data  0.037 ( 0.043)	Loss -0.2109 (-0.1182)	backward  0.603 ( 0.433)
when I run the code using horovodrun -np 1 --start-timeout 300 python run.py,  the output is
[1,0]<stdout>:Epoch: [0][    0/40036]	Time  4.635 ( 4.635)	Data  3.282 ( 3.282)	Loss 0.0035 (0.0035)	backward  0.642 ( 0.642)
[1,0]<stdout>:Epoch: [0][   50/40036]	Time 14.173 ( 9.404)	Data  0.054 ( 0.119)	Loss -0.0450 (-0.0170)	backward  0.126 ( 0.144)
[1,0]<stdout>:Epoch: [0][  100/40036]	Time 14.211 (11.006)	Data  0.054 ( 0.086)	Loss -0.0654 (-0.0339)	backward  0.123 ( 0.139)
[1,0]<stdout>:Epoch: [0][  150/40036]	Time 14.140 (11.790)	Data  0.065 ( 0.077)	Loss -0.0924 (-0.0500)	backward  0.129 ( 0.138)
[1,0]<stdout>:Epoch: [0][  200/40036]	Time 15.167 (12.465)	Data  0.070 ( 0.077)	Loss -0.0803 (-0.0619)	backward  0.132 ( 0.136)
[1,0]<stdout>:Epoch: [0][  250/40036]	Time 14.239 (12.761)	Data  0.054 ( 0.073)	Loss -0.1626 (-0.0748)	backward  0.126 ( 0.135)
[1,0]<stdout>:Epoch: [0][  300/40036]	Time 14.158 (12.961)	Data  0.061 ( 0.069)	Loss -0.1847 (-0.0881)	backward  0.122 ( 0.135)
[1,0]<stdout>:Epoch: [0][  350/40036]	Time 14.018 (13.093)	Data  0.047 ( 0.067)	Loss -0.1645 (-0.1005)	backward  0.135 ( 0.134)
[1,0]<stdout>:Epoch: [0][  400/40036]	Time 14.135 (13.209)	Data  0.054 ( 0.065)	Loss -0.2561 (-0.1133)	backward  0.123 ( 0.134)
I also try the code on three machines, master machine has 2 GPUs and every slaver machine has 1 GPU. I run the code using horovodrun -np 4 -H localhost:2,172.26.0.60:1,172.26.0.178:1 -p 1234 --start-timeout 300 python run.py, and the output is 
[1,3]<stdout>:Epoch: [0][    0/10009][1,3]<stdout>:	[1,3]<stdout>:Time  8.160 ( 8.160)[1,3]<stdout>:	[1,3]<stdout>:Data  3.247 ( 3.247)[1,3]<stdout>:	[1,3]<stdout>:Loss 0.0013 (0.0013)[1,3]<stdout>:	[1,3]<stdout>:backward  4.074 ( 4.074)
[1,2]<stdout>:Epoch: [0][    0/10009]	Time  8.194 ( 8.194)	Data  4.135 ( 4.135)	Loss 0.0027 (0.0027)	backward  3.258 ( 3.258)
[1,0]<stdout>:Epoch: [0][    0/10009]	Time  8.250 ( 8.250)	Data  5.184 ( 5.184)	Loss -0.0039 (-0.0039)	backward  1.927 ( 1.927)
[1,1]<stdout>:Epoch: [0][    0/10009]	Time  8.236 ( 8.236)	Data  4.979 ( 4.979)	Loss -0.0060 (-0.0060)	backward  2.299 ( 2.299)
[1,3]<stdout>:Epoch: [0][   50/10009]	Time 53.943 (31.051)	Data  0.006 ( 0.069)	Loss -0.0513 (-0.0230)	backward  1.041 ( 1.057)
[1,2]<stdout>:Epoch: [0][   50/10009]	Time 53.914 (31.054)	Data  0.006 ( 0.087)	Loss -0.0364 (-0.0218)	backward  1.022 ( 1.038)
[1,1]<stdout>:Epoch: [0][   50/10009]	Time 53.904 (31.070)	Data  0.005 ( 0.114)	Loss -0.0413 (-0.0216)	backward  0.943 ( 0.963)
[1,0]<stdout>:Epoch: [0][   50/10009]	Time 53.915 (31.082)	Data  0.032 ( 0.116)	Loss -0.0420 (-0.0219)	backward  0.966 ( 0.954)
[1,3]<stdout>:Epoch: [0][  100/10009]	Time 52.576 (38.226)	Data  0.005 ( 0.038)	Loss -0.0794 (-0.0411)	backward  0.929 ( 1.014)
[1,2]<stdout>:Epoch: [0][  100/10009]	Time 52.573 (38.227)	Data  0.011 ( 0.047)	Loss -0.0616 (-0.0386)	backward  0.916 ( 1.003)
[1,1]<stdout>:Epoch: [0][  100/10009]	Time 52.590 (38.243)	Data  0.018 ( 0.066)	Loss -0.1031 (-0.0410)	backward  0.912 ( 0.937)
[1,0]<stdout>:Epoch: [0][  100/10009]	Time 52.587 (38.250)	Data  0.039 ( 0.069)	Loss -0.0566 (-0.0407)	backward  0.782 ( 0.931)
[1,2]<stdout>:Epoch: [0][  150/10009]	Time 52.839 (41.880)	Data  0.011 ( 0.034)	Loss -0.0757 (-0.0542)	backward  0.941 ( 0.992)
[1,3]<stdout>:Epoch: [0][  150/10009]	Time 52.844 (41.881)	Data  0.026 ( 0.027)	Loss -0.1117 (-0.0560)	backward  0.943 ( 1.001)
[1,1]<stdout>:Epoch: [0][  150/10009]	Time 52.827 (41.889)	Data  0.010 ( 0.051)	Loss -0.1082 (-0.0556)	backward  0.918 ( 0.927)
[1,0]<stdout>:Epoch: [0][  150/10009]	Time 52.841 (41.898)	Data  0.029 ( 0.053)	Loss -0.0919 (-0.0551)	backward  0.980 ( 0.927)

I find that the backward time increases when I use more GPUs, I want know if the communication between GPUs causes the 
problem? Thank you!

more information:
horovodrun --check-build

Available Frameworks:
    [X] TensorFlow
    [X] PyTorch
    [X] MXNet

Available Controllers:
    [X] MPI
    [X] Gloo

Available Tensor Operations:
    [X] NCCL
    [ ] DDL
    [ ] CCL
    [X] MPI
    [X] Gloo

**Comment by chongxiaoc** [comment]
That is not from backward.
Extra communication overhead is from `gradients allreduce across all GPUs`, which is inside `op.step`.

When you are using a single GPU, you don't do allreduce of gradients.

**Comment by GG-Bond-007** [comment]
Thanks for your reply! I don't know what's the problem happened in backward, but backward time actually increases when more GPUs are used. Do you think if it has anything to do with horovod? Thank you!

**Comment by chongxiaoc** [comment]
@GG-Bond-007 If you you record time for `optimzer.step()` only, you will see the difference.
gradients allreduce for multi-gpus happens inside `optimzer.step()`.
